core/network/src/service.rs

```diff
@@ -589,14 +601,23 @@ fn run_thread<B: BlockT + 'static, S: NetworkSpecialization<B>, H: ExHashT>(
 
 	// Interval at which we send status updates on the `status_sinks`.
 	let mut status_interval = tokio::timer::Interval::new_interval(STATUS_INTERVAL);
+	// Interval at which we update the `connected_peers` Arc.
+	let mut connected_peers_interval = tokio::timer::Interval::new_interval(CONNECTED_PEERS_INTERVAL);
 
 	futures::future::poll_fn(move || {
 		while let Ok(Async::Ready(_)) = status_interval.poll() {
 			let status = protocol.status();
 			status_sinks.lock().retain(|sink| sink.unbounded_send(status.clone()).is_ok());
 		}
 
-		match protocol.poll(&mut Ctxt(&mut network_service.lock(), &peerset)) {
+		while let Ok(Async::Ready(_)) = connected_peers_interval.poll() {
```

Comment:
As noted with the updating of `is_offline` and `is_major_syncing`, `peers` was previously only updated when something changed, now it will be updated every interval, whether something changed or not. 

Unlike the other fields, which are updated at each `poll` of the future returned by `run_thread`(I think), the interval could actually introduce some delays in the updating in some cases. I think this potentially could introduce problems, since whoever uses `peers` could still see peers which would have already been removed from `ContextData`.  This could be made worse if anything ends-up blocking the networking runtime, since the interval could turn out to be much longer than intended. 

So if you must move everything out of protocol, perhaps also update `peers` at each turn of `run_thread`?

Reply:
Ah yeah, I went for this because peers() is only ever used to answer an RPC query for diagnostic purposes, where it really doesn't matter if you're 500ms late.

It's probably worth adding a comment about that, though.

Reply:
In the last commit I also renamed `peers()` to `peers_debug_info()` in order to make it clearer.

Reply:
Good idea.